src/Adapter/Image/Uploader/ProductImageUploader.php

```diff
@@ -0,0 +1,248 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Image\Uploader;
+
+use ErrorException;
+use Hook;
+use Image;
+use ImageManager;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\CannotUnlinkImageException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\ImageException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\ImageNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\Exception\ImageUpdateException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\ValueObject\ImageId;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageOptimizationException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageUploadException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\MemoryLimitException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\ImageUploaderInterface;
+use PrestaShopException;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+
+final class ProductImageUploader extends AbstractImageUploader implements ImageUploaderInterface
+{
+    /**
+     * @var array
+     */
+    private $contextShopIdsList;
+
+    /**
+     * @var int
+     */
+    private $contextShopId;
+
+    /**
+     * @var bool
+     */
+    private $isLegacyImageMode;
+
+    /**
+     * @param array $contextShopIdsList
+     * @param int $contextShopId
+     * @param bool $isLegacyImageMode
+     */
+    public function __construct(
+        array $contextShopIdsList,
+        int $contextShopId,
+        bool $isLegacyImageMode
+    ) {
+        $this->contextShopIdsList = $contextShopIdsList;
+        $this->contextShopId = $contextShopId;
+        $this->isLegacyImageMode = $isLegacyImageMode;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function upload($productId, UploadedFile $uploadedImage, ?ImageId $imageId = null)
+    {
+        if (!$imageId) {
+            throw new ImageUploadException('Image id is required to create path for product image');
+        }
+
+        $temporaryImageName = $this->moveToTemporaryDir($uploadedImage);
+        $image = $this->loadImageEntity($imageId);
+
+        $this->checkMemory($temporaryImageName);
+        $this->createDestinationDirectory($image);
+        $this->copyToDestination($temporaryImageName, $image);
+        $this->checkImageIsAllowedForUpload($uploadedImage);
+        $this->generateDifferentSize(
+            $productId,
+            $temporaryImageName-$this->getDestinationPath($image, false),
+            'products'
+        );
+
+        Hook::exec('actionWatermark', ['id_image' => $image->id, 'id_product' => $productId]);
+        $this->updateCover($image);
+        $image->associateTo($this->contextShopIdsList);
+
+        try {
+            //@todo: this line was originally executed before the hook 'actionWatermark'. does it matter? AdminProductsController::2881
```

Comment:
I tried to trace it back in history to understand the "necessary to avoid file hacking" comment but it's too old. I gues it's to avoid putting in tmp folder a hacker file that could be executed later.
If it does not hurt, let's keep it.

@PierreRambaud can you tell us if this is a best practice ?

Reply:
I don't really understand why they did it :thinking: 
But it's always good to remove temporary files if they are not needed.